src/Speech/AligningFeatureExtractor.cc

```diff
@@ -164,14 +187,20 @@ bool AligningFeatureExtractor::initializeAlignment() {
         error("Failed to extract alignment.");
         return false;
     }
+    if ( peakyAlignment_ )
+        makePeakyAlignment(&alignmentRef_->data());
+    else if ( forceSingleState_ )
+        makeSingleStateAlignment(&alignmentRef_->data());
```

Comment:
How "peaky-alignment" and "force-single-state" play together is not obvious without looking at the code. At the moment "peaky-alignment" overwrites "force-single-state" in that it always also reduces the alignment to single-state regardless of the value of `forceSingleState_`. I think it would be clearer if "peaky-alignment" reduces the alignment to single-state if and only if "force-single-state" is set to true, then they would have their independent effects. This would also allow using peakiness for 3-state alignments in an obvious way. The `makePeakyAlignment` function can be easily extended to preserve the states, so I would change it to a template function like
`template <bool forceSingleState> void AligningFeatureExtractor::makePeakyAlignment` and then change this case distinction to
```suggestion
    if ( peakyAlignment_ ) {
        if ( forceSingleState_ )
            makePeakyAlignment<true>(&alignmentRef_->date());
        else
            makePeakyAlignment<false>(&alignmentRef_->data());
    } else if ( forceSingleState_ )
        makeSingleStateAlignment(&alignmentRef_->data());
```

Reply:
yes, peaky alignment is not necessarily single state and these can be separated.  As  forceSingleState_ is a data member already and there's not much overhead with the for loop over alignments, a template function is not necessary here.